gobblin-utility/src/main/java/gobblin/util/io/StreamUtils.java

```diff
@@ -96,4 +109,102 @@ public static long copy(ReadableByteChannel inputChannel, WritableByteChannel ou
 
     return totalBytesRead;
   }
+
+  /**
+   * Creates a tar gzip file using a given {@link Path} as input and a given {@link Path} as a destination. If the given
+   * input is a file then only that file will be added to tarball. If it is a directory then the entire directory will
+   * be recursively put into the tarball.
+   *
+   * @param fs the {@link FileSystem} the input exists, and the the output should be written to.
+   * @param sourcePath the {@link Path} of the input files, this can either be a file or a directory.
+   * @param destPath the {@link Path} that tarball should be written to.
+   */
+  public static void tar(FileSystem fs, Path sourcePath, Path destPath)
+      throws IOException {
+    tar(fs, fs, sourcePath, destPath);
+  }
+
+  /**
+   * Similiar to {@link #tar(FileSystem, Path, Path)} except the source and destination {@link FileSystem} can be different.
+   *
+   * @see {@link #tar(FileSystem, Path, Path)}
+   */
+  public static void tar(FileSystem sourceFs, FileSystem destFs, Path sourcePath, Path destPath)
+      throws IOException {
+    try (FSDataOutputStream fsDataOutputStream = destFs.create(destPath);
+        TarArchiveOutputStream tarArchiveOutputStream = new TarArchiveOutputStream(
+            new GzipCompressorOutputStream(fsDataOutputStream), Charsets.UTF_8.name())) {
+
+      FileStatus fileStatus = sourceFs.getFileStatus(sourcePath);
+
+      if (fileStatus.isDir()) {
+        dirToTarArchiveOutputStreamRecursive(fileStatus, sourceFs, new File("."), tarArchiveOutputStream);
+      } else {
+        try (FSDataInputStream fsDataInputStream = sourceFs.open(sourcePath)) {
+          fileToTarArchiveOutputStream(fileStatus, fsDataInputStream, new File(sourcePath.getName()),
```

Comment:
`sourcePath` and `destPath` can be either HDFS or local files depending on the `FileSystem` implementation. The reason we have to make a `File` object for `sourcePath` is that `TarArchiveOutputStream` uses `File` objects to make new entries in the tarball.

e.g. `tarArchiveOutputStream.putArchiveEntry(new TarArchiveEntry(new File("."));`